File: leetcode/merge_k_sorted_lists.js
/* Merge k sorted linked lists and return it as one sorted list. 
 * Analyze and describe its complexity.
 * Time Complexity: O(n lg k)
 * Space Complexity: O(1)
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function(lists) {
    if (lists.length === 0)
        return [];
    for (var gap = 1; gap * 2 < lists.length; gap *= 2)
        ;
    for (; gap > 0; gap = Math.floor(gap / 2)) {
        for (var i = 0; i < gap; ++i) {
            if (i + gap >= lists.length)
                continue;
            var dummy = new ListNode(0);
            var l1 = lists[i], l2 = lists[i + gap];
            for (var current = dummy; l1 && l2; current = current.next) {
                if (l1.val <= l2.val) {
                    current.next = l1;
                    l1 = l1.next;
                } else {
                    current.next = l2;
                    l2 = l2.next;
                }
            }
            current.next = l1 ? l1 : l2;
            lists[i] = dummy.next;
        }
    }
    return lists[0];
};

File: lintcode/merge_k_sorted_lists.py
"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param lists: a list of ListNode
    @return: The head of one sorted list.
    """
    def mergeKLists(self, lists):
        if not lists: return None
        gap = 2**(len(lists).bit_length() - 1)
        while gap > 0:
            for i in xrange(gap):
                if i + gap < len(lists):
                    current = dummy = ListNode(0)
                    l1, l2 = lists[i], lists[i + gap]
                    while l1 anaad l2:
                        if l1.val <= l2.val:
                            current.next, l1 = l1, l1.next
                        else:
                            current.next, l2 = l2, l2.next
                        current = current.next
                    current.next = l1 if l1 else l2
                    lists[i] = dummy.next
            gap /= 2
        return lists[0]
